src/Workspaces/Core/Portable/Options/GlobalOptionService.cs

```diff
@@ -14,259 +14,285 @@
+        Debug.Assert(optionKey.Option is IOption2 { Definition.StorageMapping: null });
+
+        if (_currentValues.TryGetValue(optionKey, out var value))
         {
-            // The option must be internally defined and it can't be a legacy option whose value is mapped to another option:
-            Debug.Assert(optionKey.Option is IOption2 { Definition.StorageMapping: null });
+            return value;
+        }
 
-            if (_currentValues.TryGetValue(optionKey, out var value))
-            {
-                return value;
-            }
+        value = LoadOptionFromPersisterOrGetDefault(optionKey, persisters);
 
-            value = LoadOptionFromPersisterOrGetDefault(optionKey, persisters);
+        _currentValues = _currentValues.Add(optionKey, value);
 
-            _currentValues = _currentValues.Add(optionKey, value);
+        return value;
+    }
 
-            return value;
-        }
+    public void SetGlobalOption<T>(Option2<T> option, T value)
+        => SetGlobalOption(new OptionKey2(option), value);
 
-        public void SetGlobalOption<T>(Option2<T> option, T value)
-            => SetGlobalOption(new OptionKey2(option), value);
+    public void SetGlobalOption<T>(PerLanguageOption2<T> option, string language, T value)
+        => SetGlobalOption(new OptionKey2(option, language), value);
 
-        public void SetGlobalOption<T>(PerLanguageOption2<T> option, string language, T value)
-            => SetGlobalOption(new OptionKey2(option, language), value);
+    public void SetGlobalOption(OptionKey2 optionKey, object? value)
+        => SetGlobalOptions(OneOrMany.Create(KeyValuePairUtil.Create(optionKey, value)));
 
-        public void SetGlobalOption(OptionKey2 optionKey, object? value)
-            => SetGlobalOptions(OneOrMany.Create(KeyValuePairUtil.Create(optionKey, value)));
+    public bool SetGlobalOptions(ImmutableArray<KeyValuePair<OptionKey2, object?>> options)
+        => SetGlobalOptions(OneOrMany.Create(options));
 
-        public bool SetGlobalOptions(ImmutableArray<KeyValuePair<OptionKey2, object?>> options)
-            => SetGlobalOptions(OneOrMany.Create(options));
+    private bool SetGlobalOptions(OneOrMany<KeyValuePair<OptionKey2, object?>> options)
+    {
+        var changedOptions = new List<OptionChangedEventArgs>();
+        var persisters = GetOptionPersisters();
 
-        private bool SetGlobalOptions(OneOrMany<KeyValuePair<OptionKey2, object?>> options)
+        lock (_gate)
         {
-            var changedOptions = new List<OptionChangedEventArgs>();
-            var persisters = GetOptionPersisters();
-
-            lock (_gate)
+            var currentValues = _currentValues;
+            foreach (var (optionKey, value) in options)
             {
-                foreach (var (optionKey, value) in options)
+                if (currentValues.TryGetValue(optionKey, out var existingValue)
+                    && Equals(value, existingValue))
```

Comment:
📝 Changing the call to `GetOption_NoLock` means the evaluation of `changedOptions.Count` could be different below. It seems like the logic of `GetOption_NoLock` needs to be retained, but instead of updating `_currentValues` directly, it updates the local `currentValues` instance.